Use property manager in batch algorithm runner

gemmaguest opened the pull request:
Note that this work is by @DavidFair and @ConorMFinn rather than myself.

This PR makes changes to the BatchAlgorithmRunner and related classes to change from using AlgorithmRuntimeProps which was previously a map of string to string, to using the property manager so that we can set proper property types. This is required for upstream work in the reflectometry GUI which will need to pass workspaces as pointers rather than string names, because they will not be in the ADS.

**To test:**

Check that:
- Test something in Reflectometry GUI that relies on it still works
  - Open the ISIS Reflectometry GUI
  - Set instrument INTER. Enter run number 13460, angle 0.4
  - Click process. The row should turn green when processing completes successfully.
- Test something in Indirect (from their testing instructions) still works

Refs #32772

*This does not require release notes* because **it is an internal changes that does not affect functionality**

---

#### Reviewer ####

Please comment on the following ([full description](http://developer.mantidproject.org/ReviewingAPullRequest.html)):

##### Code Review #####

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there is GUI work does it follow the [GUI standards](http://developer.mantidproject.org/Standards/GUIStandards.html)?
- If there are changes in the release notes then do they describe the changes appropriately?

##### Functional Tests #####

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.

DavidFair committed:
Use IPropertyManager to manage Alg Props

Explicitly check that props exist when updating

This preserves existing behaviour where it would throw trying to assign,
which existing tests rely on.

DavidFair committed:
Replace new/delete with unique_ptr in PropManagerTest

Since this is the preferred modern version

DavidFair committed:
Add method to get names for properties from PropManager

This avoids callers having to manually parse this (possibly incorrectly)
and is required for some upstream work

DavidFair committed:
Update expected call count for AlgStarted in tests

Since the validation now takes place in execute, rather than when we're
copying the properties. This means from an observer POV we are now
starting the "correct" number of algorithms too.

DavidFair committed:
Address PR comments on AlgRuntimeProps changes

Various PR comments including:
- Copyright years
- Better method naming
- Some comments in code

DavidFair committed:
Add parser for AlgRuntimeProps to string

This is required to maintain compatibility with the existing
RunPresenter class in reflectometry

DavidFair committed:
Fixup incorrectly ported comparison and tests

Fixes an incorrect comparison, which implicitly relied on string
convertable types and would fail.

Fix a test where I assumed all properties were equal, whereas the
original code simply check the properties exists and we're the same for
the list passed (inclusive)

Commit:
Fix invalid props exception printing first attr twice

This would print the first element name twice incorrectly when there are
n>2 elements, as we did not correctly move the iterator on before
starting std::accumulate

gemmaguest committed:
Fix cppcheck errors with modified files

Re #32772

DavidFair commented on qt/scientific_interfaces/ISISReflectometry/test/Runs/RunsPresenterTest.h:
```diff
@@ -682,27 +684,31 @@ class RunsPresenterTest : public CxxTest::TestSuite {
     auto options = defaultLiveMonitorReductionOptions();
-    expectGetLiveDataOptions(options);
+    expectGetLiveDataOptions(
+        std::make_unique<AlgorithmRuntimeProps>(dynamic_cast<const MantidQt::API::AlgorithmRuntimeProps &>(*options)));
```
Since this is intentionally making a copy, we could just simplify this down by calling the original method twice:
`expectGetLiveDataOptions(defaultLiveMonitorReductionOptions())`

gemmaguest committed:
Remove noexcept due to OSX C++ version

Re #32772

gemmaguest committed:
Minor tidying to simplify test code

Re #32772

cailafinn reviewed (dismissed):
It looks all good to me!

gemmaguest commented:
There seems to be a problem processing in the refl GUI. I'm looking into it.

martyngigg commented on Framework/API/src/Algorithm.cpp:
```diff
@@ -2050,6 +2050,19 @@ Property *Algorithm::getPointerToPropertyOrdinal(const int &index) const {
+ * @return A vector holding strings of property names
+ */
+std::vector<std::string> Algorithm::getDeclaredPropertyNames() const noexcept {
+  std::vector<std::string> names;
```
I took a quick look as it was in the approved list for a bit. Could I have some help in understanding if this can just call through to the  implementation in the property manager, like the other methods that are implementing that interface?

gemmaguest replied:
@DavidFair is probably the best person to answer this

DavidFair replied:
That's correct, the inheritance is a bit unusual but I didn't want to start refactoring core components in a fairly big PR.

Basically as a diagram the inheritance chain looks something like:
```
                       -  - - [ PropertyManager ] 
| IPropertyManager| - |  -  - [ IAlgorithm/Algorithm ]
                      -  - -  [ PropertyManagerOwner ]
```

This means that all the (non-templated) methods are duplicated between Algorithm / PropertyManager, instead of being inherited. This can be seen in other methods such as [Algorithm::getProperties](https://github.com/mantidproject/mantid/blob/main/Framework/API/src/Algorithm.cpp#L2051) and [PropertyManager::getProperties](https://github.com/mantidproject/mantid/blob/main/Framework/Kernel/src/PropertyManager.cpp#L660)

It would be nice to fix the inheritance chain to
```
[IPropertyManager]  - - - [ PropertyManager] - [IAlgorithm / Algorithm]
                    - - - [ PropertyManagerOwner ]
```
and completely remove all the code duplication which does identical ops in 90% of cases, but this would be a core / extra attention PR if we did it

martyngigg replied:
Yes the owner thing is a little odd and I'm happy not doing a bigger change here now. My main question here really was whether this method can just do `m_properties.getDeclaredPropertyNames()` like `getProperties` does to remove duplication of the implementation?

DavidFair replied:
Ah I hadn't spotted that it has a PropertyManager (I assumed is-a), we could definitely do that 👍

gemmaguest committed:
Fix a bug where errors are thrown for valid properties

Properties were incorrectly being identified as invalid due to the inputs to std::set_difference being unsorted.

Re #32772

DavidFair committed:
Simplify method to get names for properties from PropManager

DavidFair reviewed (dismissed):
This still works following the Indirect unscripted testing instructions (as they also used AlgorithmRuntimeProps)

I'm happy with the changes Gemma did to fix the issues in the Reflectometry GUI (thanks) 
So IMO I'm happy for this work to go

gemmaguest commented:
I've done some more testing with the reflectometry GUI and am happy this is all now working (aside from the bug resolved in #32932 but that's a problem on main)

cailafinn reviewed (dismissed):
Looks good to me!

martyngigg commented:
I'm happy with the changes here but there seem to be some build failures.

gemmaguest committed:
Fix compiler error from merge conflicts

Re #32772

gemmaguest commented:
> I'm happy with the changes here but there seem to be some build failures.

Sorry, that was due to hastily pushed conflict resolutions. @DavidFair @ConorMFinn Could you re-approve?

gemmaguest commented:
@martyngigg This is ready for another look now
